proposals/NativeInt.md

```diff
@@ -0,0 +1,263 @@
+# Native-sized integers
+
+## Summary
+[summary]: #summary
+
+Language support for a native-sized signed and unsigned integer types.
+
+The motivation is for interop scenarios and for low-level libraries.
+
+## Design
+[design]: #design
+
+New contextual keywords `nint` and `nuint` represent native signed and unsigned integer types.
+The identifiers are only treated as keywords when used in a type context and when the identifier does not otherwise bind to a symbol at that program location.
+
+The types `nint` and `nuint` are essentially aliases to underlying types `System.IntPtr` and `System.UIntPtr`, where the compiler can surface additional conversions and operations for native ints.
+
+### Constants
+
+There is no direct syntax for native int literals. Explicit casts of other integral constant values can be used instead: `(nint)42`.
+
+`nint` constants are in the range [ `int.MinValue`, `int.MaxValue` ].
+
+`nuint` constants are in the range [ `uint.MinValue`, `uint.MaxValue` ].
+
+There are no `MinValue` or `MaxValue` fields on `nint` or `nuint` because, other than `nuint.MinValue`, those values cannot be emitted as constants.
+
+Constant folding is supported for all operators: { (unary)`-`, `~`, `+`, `-`, `*`, `/`, `%`, `==`, `!=`, `<`, `<=`, `>`, `>=`, `&`, `|`, `<<`, `>>` }.
+Constant folding operations are evaluated with `Int32` and `UInt32` operands rather than native ints for consistent behavior regardless of compiler platform.
```

Comment:
>are evaluated with `Int32` and `UInt32` operands [](start = 28, length = 48)

In other words, they produce different values than are produced at runtime on some runtime platforms.

Reply:
Such combinations that could produce different values should be disallowed without explicit `unchecked` context like `int.MaxValue + 1` is disallowed today for `int` constants.

Reply:
I believe the decision in the last LDM meeting was that `nint x = int.MaxValue + 1` would be illegal, even with `unchecked` due to the platform specific differences.

Instead, constant folding would only be allowed if the operation stayed within the 32-bit range limitation put forth by the compiler.

It would probably be worth explaining this limitation or elaborating on how folding works if I'm misremembering.

Reply:
Constant expressions that exceed 32 bits will be reported as errors.

---
In reply to: [330750070](https://github.com/dotnet/csharplang/pull/2833#discussion_r330750070) [](ancestors = 330750070)

Reply:
> 
> 
> Constant expressions that exceed 32 bits will be reported as errors.
> 
> In reply to: [330750070](https://github.com/dotnet/csharplang/pull/2833#discussion_r330750070) [](ancestors = 330750070)

I don't see how unary `~` could ever be valid given that constraint.

Reply:
Good catch, thanks.

---
In reply to: [386612190](https://github.com/dotnet/csharplang/pull/2833#discussion_r386612190) [](ancestors = 386612190)